Refactor `findPreviousAndNextAttributes` to use iterative pre-order traversal

Ahmad-S792 opened the pull request:
#### 0beddeef1fecd4ac9fbd0dc195698987e8a07042
<pre>
Refactor `findPreviousAndNextAttributes` to use iterative pre-order traversal
<a href="https://bugs.webkit.org/show_bug.cgi?id=265450">https://bugs.webkit.org/show_bug.cgi?id=265450</a>
<a href="https://rdar.apple.com/119183485">rdar://119183485</a>

Reviewed by Said Abou-Hallawa.

Inspired by: <a href="https://chromium.googlesource.com/chromium/blink/+/ab5b1a60d9ef3c196db20634d36c1d66c14a042f">https://chromium.googlesource.com/chromium/blink/+/ab5b1a60d9ef3c196db20634d36c1d66c14a042f</a>

This patch refactors `findPreviousAndNextAttributes` function to use the
existing pre-order traversal methods on RenderObject to implement the traversal
in an iterative fashion instead of as a recursion.

This also leverages std::tuple and renames variable from &apos;stopAfterNext&apos;
to &apos;foundLocateElement&apos; and flatten control flow with early returns.

* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::findPreviousAndNextAttributes):
(WebCore::RenderSVGText::subtreeChildWasAdded):
(WebCore::RenderSVGText::subtreeChildWillBeRemoved):

Canonical link: <a href="https://commits.webkit.org/296257@main">https://commits.webkit.org/296257@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/645668e76bed80f80c43c377aee1a7cc4b6d277c

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/107851 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/27519 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/17937 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/113065 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/58375 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/109817 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/28214 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/36062 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/81857 "Passed tests") | [⏳ 🧪 win-tests](https://ews-build.webkit.org/#/builders/Win-Tests-EWS "Waiting to run tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/110796 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/22344 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/97167 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/62283 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/21764 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/15302 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/57818 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/91695 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/15366 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/116186 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/34920 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/25744 "Build is in progress. Recent messages:OS: Sequoia (15.5), Xcode: 16.4; Skipping applying patch since patch_id isn't provided; Checked out pull request; Skipped layout-tests; Running layout-tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/90886 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/35296 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/93445 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/90677 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/23133 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/35575 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/13336 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/30670 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/34824 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/40376 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/34566 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/37927 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/36228 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -109,34 +109,30 @@ static inline void collectLayoutAttributes(RenderObject* text, Vector<SVGTextLay
-static inline bool findPreviousAndNextAttributes(RenderElement& start, RenderSVGInlineText* locateElement, bool& stopAfterNext, SVGTextLayoutAttributes*& previous, SVGTextLayoutAttributes*& next)
+static inline bool findPreviousAndNextAttributes(RenderElement& root, RenderSVGInlineText* locateElement, SVGTextLayoutAttributes*& previous, SVGTextLayoutAttributes*& next)
 {
     ASSERT(locateElement);
```
`locateElement` can be `RenderSVGInlineText&` so this assertion can be removed.

Ahmad-S792 replied:
Done.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -109,34 +109,30 @@ static inline void collectLayoutAttributes(RenderObject* text, Vector<SVGTextLay
-    // FIXME: Make this iterative.
-    for (CheckedRef child : childrenOfType<RenderObject>(start)) {
-        if (auto* text = dynamicDowncast<RenderSVGInlineText>(child.get())) {
+    bool stopAfterNext = false;
```
The name `stopAfterNext` is a little bit confusing. What does `Next` mean in this context? 

I would suggest renaming it to `foundLocateElement` or something. `stopAfterNext` is set to true only when `locateElement == text` is true.

Ahmad-S792 replied:
Done.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -109,34 +109,30 @@ static inline void collectLayoutAttributes(RenderObject* text, Vector<SVGTextLay
-        if (auto* text = dynamicDowncast<RenderSVGInlineText>(child.get())) {
+    bool stopAfterNext = false;
+    CheckedPtr current = root.firstChild();
+    while (current) {
```
These can be merged in one for statement:
```
for (CheckedPtr current = root.firstChild(); current; )
```

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -109,34 +109,30 @@ static inline void collectLayoutAttributes(RenderObject* text, Vector<SVGTextLay
 }
 
-static inline bool findPreviousAndNextAttributes(RenderElement& start, RenderSVGInlineText* locateElement, bool& stopAfterNext, SVGTextLayoutAttributes*& previous, SVGTextLayoutAttributes*& next)
+static inline bool findPreviousAndNextAttributes(RenderElement& root, RenderSVGInlineText* locateElement, SVGTextLayoutAttributes*& previous, SVGTextLayoutAttributes*& next)
```
The callers of this function do not check the return bool value. They have also to define and initialize the arguments `previous` and `next`. How about this prototype:
```
static inline std::tuple<SVGTextLayoutAttributes*, SVGTextLayoutAttributes*> findPreviousAndNextAttributes(RenderElement& root, RenderSVGInlineText& locateElement)
```
And the callers can just call it this way:
```
auto [previous, next] = findPreviousAndNextAttributes(*this, attributes->context());
```

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -252,11 +247,10 @@ void RenderSVGText::subtreeChildWillBeRemoved(RenderObject* child, Vector<SVGTex
     SVGTextLayoutAttributes* next = nullptr;
     if (!renderTreeBeingDestroyed())
-        findPreviousAndNextAttributes(*this, &text, stopAfterNext, previous, next);
+        std::tie(previous, next) = findPreviousAndNextAttributes(*this, text);
```
I think you can define `previous` and `next` only `if (!renderTreeBeingDestroyed())` is true
```
    if (!renderTreeBeingDestroyed()) {
        auto [previous, next] = findPreviousAndNextAttributes(...);
        if (previous)
            ...  
    }
```

Ahmad-S792 replied:
I think - I manage to capture it in this commit, if you don't mind checking.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -123,35 +124,32 @@ static inline void collectLayoutAttributes(RenderObject* text, Vector<SVGTextLay
-            if (locateElement != text) {
-                if (stopAfterNext) {
+    SVGTextLayoutAttributes* previous = nullptr;
+    SVGTextLayoutAttributes* next = nullptr;
```
No need to define `next`. It is only used twice in this function:

```
    next = text->layoutAttributes();
    return { previous, next };
```
And the end of the function:
```
    return { previous, next };
```
The one at the end of the function has to be nullptr.

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -123,35 +124,32 @@ static inline void collectLayoutAttributes(RenderObject* text, Vector<SVGTextLay
+                    return { previous, next };
                 }
-
                 previous = text->layoutAttributes();
```
Can this is be clearer?
```
if (foundLocateElement)
    return { previous, text->layoutAttributes() };

if (&locateElement == text)
    foundLocateElement = true;
else
    previous = text->layoutAttributes();
```

shallawa commented on Source/WebCore/rendering/svg/RenderSVGText.cpp:
```diff
@@ -123,35 +124,32 @@ static inline void collectLayoutAttributes(RenderObject* text, Vector<SVGTextLay
                 previous = text->layoutAttributes();
+            } else
+                foundLocateElement = true;
+        } else if (auto* childSVGInline = dynamicDowncast<RenderSVGInline>(*current)) {
```
I think if we start by checking `RenderSVGInline` this function will be clearer:
```
for (CheckedPtr current = root.firstChild(); current;) {
    if (auto* inline = dynamicDowncast<RenderSVGInline>(*current)) {
        if (auto* child = inline->firstChild())
            current = child;
        else
            current = current->nextInPreOrderAfterChildren(&root);     
        continue;
    }

    if (auto* text = dynamicDowncast<RenderSVGInlineText>(*current)) {
        ...
    }

    current = current->nextInPreOrderAfterChildren(&root);
}

Commit:
Refactor `findPreviousAndNextAttributes` to use iterative pre-order traversal
https://bugs.webkit.org/show_bug.cgi?id=265450
rdar://119183485

Reviewed by Said Abou-Hallawa.

Inspired by: https://chromium.googlesource.com/chromium/blink/+/ab5b1a60d9ef3c196db20634d36c1d66c14a042f

This patch refactors `findPreviousAndNextAttributes` function to use the
existing pre-order traversal methods on RenderObject to implement the traversal
in an iterative fashion instead of as a recursion.

This also leverages std::tuple and renames variable from 'stopAfterNext'
to 'foundLocateElement' and flatten control flow with early returns.

* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::findPreviousAndNextAttributes):
(WebCore::RenderSVGText::subtreeChildWasAdded):
(WebCore::RenderSVGText::subtreeChildWillBeRemoved):

Canonical link: https://commits.webkit.org/296257@main
